core/src/main/java/amino/run/oms/migrationdecision/MigrationPredictor.java

```diff
@@ -0,0 +1,218 @@
+
+import amino.run.common.MicroServiceNotFoundException;
+import amino.run.common.ReplicaID;
+import amino.run.kernel.common.GlobalKernelReferences;
+import amino.run.kernel.common.KernelServerNotFoundException;
+import amino.run.kernel.metric.NodeMetric;
+import amino.run.kernel.metric.RPCMetric;
+import amino.run.oms.KernelServerManager;
+import amino.run.oms.MicroServiceManager;
+import java.net.InetSocketAddress;
+import java.rmi.RemoteException;
+import java.util.*;
+import java.util.logging.Logger;
+
+/**
+ * Class used for predicting processing time for a replica.
+ *
+ * @author AmitRoushan
+ */
+public class MigrationPredictor {
+    private MicroServiceManager microServiceManager;
+    private KernelServerManager serverManager;
+    private Logger logger = Logger.getLogger(MigrationPredictor.class.getName());
+    private TreeSet<AbstractMap.SimpleEntry<InetSocketAddress, Long>> predictedTime;
+    private final int TIME_OPTIMIZATION_THRESHOLD = 10; // in percentage
+
+    public MigrationPredictor(
+            KernelServerManager serverManager, MicroServiceManager microServiceManager) {
+        this.serverManager = serverManager;
+        this.microServiceManager = microServiceManager;
+        predictedTime =
+                new TreeSet<AbstractMap.SimpleEntry<InetSocketAddress, Long>>(
+                        new Comparator<AbstractMap.SimpleEntry<InetSocketAddress, Long>>() {
+                            @Override
+                            public int compare(
+                                    AbstractMap.SimpleEntry<InetSocketAddress, Long> o1,
+                                    AbstractMap.SimpleEntry<InetSocketAddress, Long> o2) {
+                                if (o1.getValue() < o2.getValue()) return 1;
+                                else if (o1.getValue() > o2.getValue()) return -1;
+                                return 0;
+                            }
+                        });
+    }
+
+    /**
+     * Predict best kernel server for replica by constructing heap based on time optimization on
+     * available kernel server
+     *
+     * @param replicaID
+     * @param currentKernelServer
+     * @return kernel server address
+     */
+    public InetSocketAddress getBestKernelServer(
+            ReplicaID replicaID, InetSocketAddress currentKernelServer) {
+        // get all kernel server
+        try {
+            List<InetSocketAddress> kernelServers =
+                    GlobalKernelReferences.nodeServer.oms.getServers(null);
+
+            // predict time optimization for replica on each kernel server and push it into sorted
+            // heap
+            for (InetSocketAddress ks : kernelServers) {
+                predictedTime.add(
+                        new AbstractMap.SimpleEntry<InetSocketAddress, Long>(
+                                ks, getOptimizedTime(replicaID, currentKernelServer, ks)));
+            }
+        } catch (RemoteException e) {
```

Comment:
Don;'t we need to throw these exceptions for the caller to handle, rather than just log them here?

Reply:
Oh, I see, if an exception occurs, you just return the current node as the best node.  Fair enough.  It seems better to catch this exception higher up the call stanc, and explicitly avoid moving the replica in that code instead.